executor/calibrate_resource.go

```diff
@@ -83,13 +91,72 @@ type baseResourceCost struct {
 	writeReqCount uint64
 }
 
+const (
+	valuableUsageThreshold = 0.2
+	lowUsageThreshold      = 0.1
+	percentOfPass          = 0.9
+	discardRate            = 0.1
+
+	maxDuration = time.Hour * 24
+	minDuration = time.Minute * 10
+)
+
 type calibrateResourceExec struct {
 	baseExecutor
-
+	optionList   []*ast.DynamicCalibrateResourceOption
 	workloadType ast.CalibrateResourceType
 	done         bool
 }
 
+func (e *calibrateResourceExec) parseCalibrateDuration() (startTime time.Time, endTime time.Time, err error) {
+	var dur time.Duration
+	var ts uint64
+	for _, op := range e.optionList {
+		switch op.Tp {
+		case ast.CalibrateStartTime:
+			ts, err = staleread.CalculateAsOfTsExpr(e.ctx, op.Ts)
+			if err != nil {
+				return
+			}
+			startTime = oracle.GetTimeFromTS(ts)
+		case ast.CalibrateEndTime:
+			ts, err = staleread.CalculateAsOfTsExpr(e.ctx, op.Ts)
+			if err != nil {
+				return
+			}
+			endTime = oracle.GetTimeFromTS(ts)
+		case ast.CalibrateDuration:
+			dur, err = duration.ParseDuration(op.StrValue)
+			if err != nil {
+				return
+			}
+		}
+	}
+	if startTime.IsZero() {
+		err = errors.Errorf("start time should not be 0")
+		return
+	}
+	// If endTime is set, duration will be ignored.
+	if endTime.IsZero() {
+		if dur != time.Duration(0) {
+			endTime = startTime.Add(dur)
+		} else {
+			endTime = time.Now()
+		}
+	}
+	// check the duration
+	dur = endTime.Sub(startTime)
+	if dur > maxDuration {
+		err = errors.Errorf("the duration of calibration is too long, should be less than %s", maxDuration.String())
```

Comment:
Suggested msg:
"The duration is too long/short, which could lead to inaccurate output. Please make the duration between s% and s%. "